### PR TITLE
cache fixes

### PR DESCRIPTION
**Pull request by shimish2**
- dedupe: stat blob path before creating link

### CODE REVIEW

File: pkg/storage/storage.go [hunk 1]
```diff
@@ -1084,6 +1091,13 @@ func (is *ImageStore) checkCacheBlob(digest string) (string, error) {
 		return "", err
 	}
 
+	// there is an entry of digest in cache with empty value.
+	if dstRecord == "" {
```
Comment (rchincha) [inline pkg/storage/storage.go:1095]: empty record means "no entry found", would enforce that check in cache.PutBlob() that we don't ever write an empty record

Comment (shimish2) [reply pkg/storage/storage.go:1095]: Yes we can add additional check there also

File: errors/errors.go [hunk 2]
```diff
@@ -39,4 +39,5 @@ var (
 	ErrDuplicateConfigName     = errors.New("cli: cli config name already added")
 	ErrInvalidRoute            = errors.New("routes: invalid route prefix")
 	ErrImgStoreNotFound        = errors.New("routes: image store not found corresponding to given route")
+	ErrEmptyPath               = errors.New("cache: empty path")
```
Comment (rchincha) [inline errors/errors.go:42]: Call this ErrEmptyValue = errors.New("cache: empty value")

File: pkg/storage/cache.go [hunk 3]
```diff
@@ -51,6 +51,12 @@ func NewCache(rootDir string, name string, log zlog.Logger) *Cache {
 }
 
 func (c *Cache) PutBlob(digest string, path string) error {
+	if path == "" {
+		c.log.Error().Err(errors.ErrEmptyPath).Msg("empty path provided")
```
Comment (rchincha) [inline pkg/storage/cache.go:55]: Include the "digest" in this log

File: pkg/storage/storage.go [hunk 4]
```diff
@@ -985,7 +985,18 @@ retry:
 			}
 			goto retry
 		}
+
 		dstFi, err := os.Stat(dst)
+		if err == nil {
+			is.log.Debug().Str("blobPath", dst).Msg("dedupe: blob exists")
+			err = os.Remove(src)
```
Comment (rchincha) [inline pkg/storage/storage.go:992]: Why is this needed?

**Review by rchincha — APPROVED** [review]
lgtm
